typegen/src/typegen/error.rs

```diff
@@ -68,3 +73,31 @@ pub enum TypeSubstitutionErrorKind {
     #[error("Cannot find matching param on 'from' type.")]
     NoMatchingFromType,
 }
+
+/// Error attempting to do type substitution.
+#[derive(Debug, thiserror::Error, Default, PartialEq, Eq)]
+pub struct SettingsValidationError {
+    /// Cannot add derive for type that is not present in the registry.
+    pub derives_for_unknown_types: Vec<(syn::Path, HashSet<syn::Path>)>,
+    /// Cannot add attribute for type that is not present in the registry.
+    pub attributes_for_unknown_types: Vec<(syn::Path, HashSet<syn::Attribute>)>,
+    /// Invalid path to replace: the type getting substituted is not present in the registry.
+    pub substitutes_for_unknown_types: Vec<(syn::Path, syn::Path)>,
+}
+
+impl SettingsValidationError {
+    pub(crate) fn is_empty(&self) -> bool {
+        self.derives_for_unknown_types.is_empty()
+            && self.attributes_for_unknown_types.is_empty()
+            && self.substitutes_for_unknown_types.is_empty()
+    }
+}
+
+impl std::fmt::Display for SettingsValidationError {
+    fn fmt(
+        &self,
+        f: &mut scale_info::prelude::fmt::Formatter<'_>,
+    ) -> scale_info::prelude::fmt::Result {
+        f.write_fmt(format_args!("{:?}", self))
+    }
+}
```

Comment:
I wonder if this could be replaced with the thiserror attr on the struct:

```rust
#[error = "{:?}"]
```

But moreover, I think that a proper `Display` impl should do better than just debug printing :)